flask_mongorest/__init__.py

```diff
@@ -2,16 +2,64 @@
 from flask_mongorest.methods import Create, BulkUpdate, List
 
 
+class _DelayedApp(object):
+
+    '''
+    Stores URL rules for later merging with application URL map.
+
+    '''
+
+    def __init__(self):
+        self.url_rules = []
+
+    def add_url_rule(self, *args, **kwargs):
+        self.url_rules.append((args, kwargs))
+
+
 class MongoRest(object):
-    def __init__(self, app, **kwargs):
-        self.app = app
+
+    def __init__(self, app=None, **kwargs):
+        '''
+        Takes optional Flask application instance. If supplied, `init_app` will be
+        called to update application url map.
+
+        '''
+
         self.url_prefix = kwargs.pop('url_prefix', '')
-        app.register_blueprint(Blueprint(self.url_prefix, __name__, template_folder='templates'))
+        self.template_folder = kwargs.pop('template_folder', 'templates')
+        if app is not None:
+            self.init_app(app, **kwargs)
+        else:
+            self.app = _DelayedApp()
+
+    def init_app(self, app):
+        '''
+        Provides delayed application instance initialization to support
+        Flask application factory pattern. For further details on application
+        factories see:
+
+        http://flask.pocoo.org/docs/dev/patterns/appfactories/
+
+        and
+
+        http://mattupstate.com/python/2013/06/26/how-i-structure-my-flask-applications.html
+
+        '''
+        app.register_blueprint(
+            Blueprint(self.url_prefix,
+                      __name__,
+                      template_folder=self.template_folder))
+        if hasattr(self, 'app'):
+            if isinstance(self.app, _DelayedApp):
+                for args, kwargs in self.app.url_rules:
+                    app.add_url_rule(*args, **kwargs)
+        self.app = app
```

Comment:
Would it be better to *always* set `self.app` in `__init__` - either to `_DelayedApp()` or to a real Flask app object? We could then skip some of the ugly hasattr conditions.